operator/buffer/memory.go

```diff
@@ -29,36 +31,135 @@ func NewMemoryBufferConfig() *MemoryBufferConfig {
-	return &MemoryBuffer{}, nil
+// Build builds a MemoryBufferConfig into a Buffer
+func (c MemoryBufferConfig) Build(_ operator.BuildContext, pluginID string) (Buffer, error) {
+	return &MemoryBuffer{
+		pluginID:      pluginID,
+		buf:           make(chan *entry.Entry, c.MaxEntries),
+		maxChunkDelay: c.MaxChunkDelay.Raw(),
+		maxChunkSize:  c.MaxChunkSize,
+		closed:        false,
+	}, nil
 }
 
-// MemoryBuffer is a buffer that holds all entries in memory until Close() is called,
-// at which point it saves the entries into a database. It provides no guarantees about
-// lost entries if shut down uncleanly.
+// MemoryBuffer is a buffer that holds all entries in memory until Close() is called.
+// Once close is called all entries will be lost it is reccommended to call Drain before Close.
 type MemoryBuffer struct {
+	pluginID      string
+	buf           chan *entry.Entry
+	maxChunkDelay time.Duration
+	maxChunkSize  uint
+	closed        bool
+
+	// readLock ensures only one concurrent read/drain can process at a time.
+	// If multiple reads or a drain happen while another is processing it may cause
+	// an indefinite block or longer blocks to maxChunkDelay.
+	readLock sync.Mutex
 }
 
 // Add adds an entry onto the buffer.
 // Is a blocking call if the buffer is full
 func (m *MemoryBuffer) Add(ctx context.Context, e *entry.Entry) error {
-	return nil
+	// If buffer is closed don't allow this operation
+	if m.closed {
+		return ErrBufferedClosed
+	}
+
+	// Insert into buffer or error if context finishes before we can
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("ctx error adding to buffer: %w", ctx.Err())
+	case m.buf <- e:
+		return nil
+	}
 }
 
 // Read reads from the buffer.
 // Read will block until the there are MaxChunkSize entries or we have block as long as MachChunkDelay.
 func (m *MemoryBuffer) Read(ctx context.Context) ([]*entry.Entry, error) {
-	return nil, nil
+	// If buffer is closed don't allow this operation
+	if m.closed {
+		return nil, ErrBufferedClosed
+	}
+
+	m.readLock.Lock()
+	defer m.readLock.Unlock()
+
+	entries := make([]*entry.Entry, 0, m.maxChunkSize)
+
+	// Time for max Chunk delay
+	// Using a timer here rather than a context with deadline as we don't want to confuse a context timeout with a maxChunkDelay timeout
+	timer := time.NewTimer(m.maxChunkDelay)
+	defer timer.Stop()
+
+LOOP:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("ctx error during buffer read: %w", ctx.Err())
+		case <-timer.C:
+			// Timer has reached maxChunkDelay break and return
+			break LOOP
+		case e := <-m.buf:
+			entries = append(entries, e)
+
+			// If we've reached the maxChunkSize break and return
+			if len(entries) == int(m.maxChunkSize) {
+				break LOOP
+			}
+		}
+	}
+
+	return entries, nil
 }
 
-// Drain drains all contents currently in the buffer to the returned entry
+// Drain drains all contents currently in the buffer at the time of calling Drain.
```

Comment:
How do you envision this being used by the output operator? I'm guessing it would have to do its own chunking or splitting after calling drain and getting a larger than normal slice back?

Reply:
So Drain ideally is to be called right before Close, mainly for the memory buffer so we don't loose everything in the buffer. Nothing stops you from calling it at any time then it's up to you to process. The operators have no concept of chunking anyway so really shouldn't matter to the consumer.